design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,195 @@
+|-------------------|--------------------------------------------------------------------------------------------------------------------------------------------|
+| 200 OK            | Full indication of the injection status for the client. 200 states that the cert is injected, and 401 states that the cert is not injected |
+| 202 Accepted      | Fast response so the client is not hanging and doesn't get to its timeout in case the injection takes too long                             |
+
+It's worth mentioning that we don't really have a backwards-compatibility issue as also today we return a 200 OK response before the cert is fully injected. Even before the wait that we added to client, it read the cert and failed - before retrying. So  changing the response to 202 Accepted is backwards-compatible.
+
+#### Investigation
+1. Is there a risk in waiting for the cert to be injected before returning a response?
+1. If we choose to return 202 Accepted, will we tell the client that the injection succeeded/failed? If so, how?
+
+#### Decision
+
+ Will be decided upon investigation.
+
+### Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+TODO: Add diagram once we decide on a solution
+
+### Backwards compatibility
+[//]: # "Address how you are going to handle backwards compatibility, if necessary"
+At current, the authn-client sends the "inject-client-cert" request and then  tries to read the cert. As this solution will just fix the way that we inject the  cert, we should have any backwards-compatibility issues.
+
+The only solution that will introduce a backwards-compatibility issue is if the  authn-client will change its State to indicate to the server that the cert is installed but i don't see any value in this option. 
+
+### Performance
+[//]: # "Elaborate on whether your solution will affect the product's performance, and how"
+The performance of the "inject-client-cert" request will be affected by the decision of [this section](#inject-client-cert-response-code).
+
+### Affected Components
+[//]: # "Address the components that will be affected by your solution [Conjur, DAP, clients, integrations, etc.]"
+- Conjur
+- DAP
+- conjur-authn-k8s-client
+- secrets-provider-for-k8s
+
+### Cross-team dependencies
+[//]: # "Elaborate about your solution's requirements or needs from other teams"
+We would like to get a review on the solution from @cyberark/conjur-core-team  and @cyberark/community-and-integrations-team. There are no requirements from  the Infra team.
+
+## Security
+[//]: # "Are there any security issues with your solution? Even if you mentioned them somewhere in the doc it may be convenient for the security architect review to have them centralized here"
+1. Is it ok to pass the base64 encoded cert in the URL? Is it less secure than sending it in the STDIN?
```

Comment:
While the traffic is fully encrypted when using https, including the URL. It's still possible that the URL will be visible in certain locations, while the request travels to its destination. See this [page](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url) for more details.

It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.

Reply:
So while I agree with @Tovli that option 2 looks the most promising technically. Let's make sure that it's safe enough to use.

Reply:
@sashaCher - would appreciate your response here. Mainly to understand the severity of having option 2 with follower outside the cluster

Reply:
I'm totally agree with @rafis3.
Despite HTTPS standard guarantees that url is encrypted (except its FQDN part) we have no guarantee that it's not logged in any of upstream systems. For example in conjur case nginx logs all urls.

Reply:
@sashaCher @rafis3 so do you see a security concern in this option? Do you suggest not to use it or do we have something to further investigate?

Reply:
My suggestion is not to use this option due to security concerns.

Reply:
Maybe it is possible to pass certificate as a body of POST request instead of URL ? It will not have a security problem then.
@sashaCher - WDYT ? @orenbm - is it possible ?

Reply:
i may be missing something but isn't sending data via STDIN the way to send a "body" in a websocket?

can you elaborate on passing a body in a websocket?

Reply:
@oburstein-hub, this connection is using websocket.
This means there's no "body".
The server treats as STDIN any bytes received on the socket after it is opened.
@orenbm suggests passing the certificate in the URL instead of sending it on the socket.

Reply:
I found syntax like this: WebSocket.send("Hello server!");
Instead of "Hello server!" can we send certificate ?

Reply:
yes, this is [what we do today](https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/kubectl_exec.rb#L60).

Reply:
thanks @sashaCher and @rafis3 . I added this paragraph to Option 2:
```
We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.

Thus, we will not use this option due to security concerns.
```